pkg/chains/storage/grafeas/grafeas_test.go

```diff
@@ -0,0 +1,382 @@
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+
+	attestationpb "github.com/grafeas/grafeas/proto/v1beta1/attestation_go_proto"
+	commonpb "github.com/grafeas/grafeas/proto/v1beta1/common_go_proto"
+	pb "github.com/grafeas/grafeas/proto/v1beta1/grafeas_go_proto"
+	"github.com/tektoncd/chains/pkg/config"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	gstatus "google.golang.org/grpc/status"
+	"google.golang.org/protobuf/testing/protocmp"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	logtesting "knative.dev/pkg/logging/testing"
+)
+
+type args struct {
+	tr        *v1beta1.TaskRun
+	payload   []byte
+	signature string
+	opts      config.StorageOpts
+}
+
+type testConfig struct {
+	name    string
+	args    args
+	wantErr bool
+}
+
+type mockGrafeasV1Beta1Server struct {
+	// Embed for forward compatibility.
+	// Tests will keep working if more methods are added in the future.
+	pb.UnimplementedGrafeasV1Beta1Server
+
+	// Assume there is only one project for storing notes and occurences
+	occurences map[string]*pb.Occurrence
+	notes      map[string]*pb.Note
+}
+
+var grafeasServer mockGrafeasV1Beta1Server
+
+func TestBackend_ServerConnCheck(t *testing.T) {
+	tests := []struct {
+		serverName string
+		wantErr    bool
+	}{
+		{serverName: "dns:///containeranalysis.googleapis.com", wantErr: false},
+		{serverName: "dns:///containeranalysis.fake.com", wantErr: true},
+	}
+
+	for _, test := range tests {
+		if err := checkTrustedHost(test.serverName); (err != nil) != test.wantErr {
+			t.Errorf("The behaviour of checking trusted host is wrong. error = %v, wantErr = %v", err, test.wantErr)
+		}
+	}
+}
+
+func TestBackend_StorePayload(t *testing.T) {
+	tests := []testConfig{
+		{
+			name: "TEST 1: intoto for taskrun, no error",
+			args: args{
+				tr: &v1beta1.TaskRun{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "foo1",
+						Name:      "bar1",
+						UID:       types.UID("uid1"),
+					},
+				},
+				payload:   []byte("taskrun payload"),
+				signature: "taskrun signature",
+				opts:      config.StorageOpts{Key: "taskrun.uuid", PayloadFormat: formats.PayloadTypeInTotoIte6},
+			},
+			wantErr: false,
+		},
+		{
+			name: "TEST 2: simplesining for oci, no error",
+			args: args{
+				tr: &v1beta1.TaskRun{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: "foo2",
+						Name:      "bar2",
+						UID:       types.UID("uid2"),
+					},
+					Status: v1beta1.TaskRunStatus{
+						TaskRunStatusFields: v1beta1.TaskRunStatusFields{
+							TaskRunResults: []v1beta1.TaskRunResult{
+								{Name: "IMAGE_DIGEST", Value: "sha256:cfe4f0bf41c80609214f9b8ec0408b1afb28b3ced343b944aaa05d47caba3e00"},
+								{Name: "IMAGE_URL", Value: "gcr.io/test/kaniko-chains1"},
+							},
+						},
+					},
+				},
+				payload:   []byte("oci payload"),
+				signature: "oci signature",
+				// the Key field must be the same as the first 12 chars of the image digest
```

Comment:
What happens if it's not? Is this something we should add additional validation for?

Reply:
Thanks for the comment! In the implementation code of `chains.SignTaskRun` function, the key is set, and for oci image, it will be set to the first 12 chars of the image digest. So in this test, we just want to add this information here in case others might need to know in order to add more tests later.

I added more comments here in the latest version.